pkg/provisioner/filestorage/filestorage.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestorage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	"github.com/oracle/oci-go-sdk/common"
+	"github.com/oracle/oci-go-sdk/core"
+	"github.com/oracle/oci-go-sdk/filestorage"
+	"github.com/oracle/oci-go-sdk/identity"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/pkg/api/v1"
+
+	"github.com/oracle/oci-volume-provisioner/pkg/oci/client"
+	"github.com/oracle/oci-volume-provisioner/pkg/provisioner/plugin"
+)
+
+const (
+	ociVolumeID            = "ociVolumeID"
+	ociExportID            = "ociExportID"
+	volumePrefixEnvVarName = "OCI_VOLUME_NAME_PREFIX"
+	fsType                 = "fsType"
+	subnetID               = "subnetId"
+	mntTargetID            = "mntTargetId"
+)
+
+// filesystemProvisioner is the internal provisioner for OCI filesystem volumes
+type filesystemProvisioner struct {
+	client client.ProvisionerClient
+}
+
+var _ plugin.ProvisionerPlugin = &filesystemProvisioner{}
+
+// NewFilesystemProvisioner creates a new file system provisioner that creates
+// filsystems using OCI file system service.
+func NewFilesystemProvisioner(client client.ProvisionerClient) plugin.ProvisionerPlugin {
+	return &filesystemProvisioner{
+		client: client,
+	}
+}
+
+// getMountTargetFromID retrieves mountTarget from given mountTargetID
+func getMountTargetFromID(ctx context.Context, mountTargetID string, fileStorageClient client.FileStorage) *filestorage.MountTarget {
+	responseMnt, err := fileStorageClient.GetMountTarget(ctx, filestorage.GetMountTargetRequest{
+		MountTargetId: common.String(mountTargetID),
+	})
+	if err != nil {
+		glog.Errorf("Failed to retrieve mount point: %s", err)
+		return nil
+	}
+	return &responseMnt.MountTarget
+}
+
+func (filesystem *filesystemProvisioner) Provision(
+	options controller.VolumeOptions,
+	availabilityDomain *identity.AvailabilityDomain) (*v1.PersistentVolume, error) {
+
+	ctx, cancel := context.WithTimeout(filesystem.client.Context(), filesystem.client.Timeout())
```

Comment:
I know this is an existing pattern, however, it's not one I particularly like. I think we should just be able to apply a default timeout to the SDK client and avoid all this boilerplate.

Reply:
Will create an issue for this and fix this separately

Reply:
https://github.com/oracle/oci-volume-provisioner/issues/158